### PR TITLE
✨ Add user instructions and regex option

### LINKED ISSUE #142: User name pattern instructions
It would be great to be able to customise the information for completing the registration.

After the user clicks the link of the email, it to we good to tell:

Please use a user name of schema: "name.surname"

Thanks for considering.

### PR DESCRIPTION
**Pull request by pxlfrk**
Hey,

in our organization we use the registration app. As we want to inform our users about our name schema for the usernames (see #142 ) and also check (or force) the schema through an regex-function we added the code manually to the plugin.

These Changes get's reverted every update of the App, so i decided to create this PR & extended the code to be able to configure the parameter through the backend. I also implemented an option to add a text which is embedded in the verification email (see #243 ).

### Username Policy

Through the Settings panel a regular expression can be configured. If there is a regex configured the usernames get's validatet through the function and if the test fails, the user is prompted with the generic "please choose an appropiate username" error. I also implemented a syntax check in the settings panel, but i'm not sure how to deal with incorrect inputs. Looking forward to your input here.

| Screenshot | Description |
| --- | --- |
| image | Regular expression field in the settings |
| image | Error shown if the test fails |

### User instructions
Via the Settings panel the user can configure instructions for the user with additional informations. They will be shown to the user while creating his user account and/or embedded in the verification email.

| Screenshot | Description |
| --- | --- |
| image | settings parameter |
| image | user notification in the user-creation-screen |
| image) | embedded body-text in the verification email |

Looking forward to your review & feedback! 👍 

Fix #142

### CODE REVIEW

File: lib/Controller/SettingsController.php [hunk 1]
```diff
@@ -64,6 +70,30 @@ public function admin(string $registered_user_group,
...
+			$this->config->setAppValue($this->appName, 'email_verification_hint', $email_verification_hint);
+		}
+
+		//handle regex
+		if (($username_policy_regex === '') || ($username_policy_regex === null)) {
+			$this->config->deleteAppValue($this->appName, 'username_policy_regex');
+		} elseif ((@preg_match($username_policy_regex, null) === false)) {
+			// validate regex
+			// TODO: Create SettingsException Class, throw it & show it to user
+
```
Comment (nickvergessen) [inline lib/Controller/SettingsController.php:92]:
```suggestion
			return new DataResponse([
				'data' => [
					'message' => $this->l10n->t('Invalid username policy regex'),
				],
				'status' => 'error',
			], Http::STATUS_BAD_REQUEST);
```

File: lib/Service/MailService.php [hunk 2]
```diff
@@ -102,6 +107,14 @@ public function sendTokenByMail(Registration $registration): void {
...
 			$body
 		);
+		
+		// if the parameter is set through the settings panel add to body text
+		$email_verification_hint = $this->config->getAppValue('registration', 'email_verification_hint');
+ 		if (($email_verification_hint === '') || ($email_verification_hint === null)) {
+			// do nothing if the parameter isn't set
+			} else {
+				$template->addBodyText($email_verification_hint);
+		};
```
Comment (nickvergessen) [inline lib/Service/MailService.php:117]:
```suggestion
 		if (!empty($email_verification_hint)) {
				$template->addBodyText($email_verification_hint);
		}; 
```

File: templates/admin.php [hunk 3]
```diff
@@ -59,6 +59,29 @@
 		<label for="email_is_login"><?php p($l->t('Force email as login name')); ?></label>
 	</p>
 
+	<h3><?php p($l->t('Username Policy')); ?></h3>
+	<p>
+		<label>
+			<input type="text" id="username_policy_regex" name="username_policy_regex" value="<?php p($_['username_policy_regex']);?>" placeholder="/^[a-z-]+\.[a-z-]+$/">
```
Comment (nickvergessen) [inline templates/admin.php:65]:
```suggestion
			<input type="text" id="username_policy_regex" name="username_policy_regex" value="<?php p($_['username_policy_regex']);?>" placeholder="E.g.: /^[a-z-]+\.[a-z-]+$/">
```

File: templates/admin.php [hunk 4]
```diff
@@ -59,6 +59,29 @@
 		<label for="email_is_login"><?php p($l->t('Force email as login name')); ?></label>
 	</p>
 
+	<h3><?php p($l->t('Username Policy')); ?></h3>
```
Comment (nickvergessen) [inline templates/admin.php:62]:
```suggestion
	<h3><?php p($l->t('Username policy')); ?></h3>
```

File: templates/admin.php [hunk 5]
```diff
@@ -59,6 +59,29 @@
...
+	<p>
+		<label>
+			<input type="text" id="username_policy_regex" name="username_policy_regex" value="<?php p($_['username_policy_regex']);?>" placeholder="/^[a-z-]+\.[a-z-]+$/">
+		</label>
+	</p>
+	<em><?php p($l->t('If configured usernames will be validated through the regular expression. If the validation fails the user is prompted with a generic error. Make sure your regex is working correctly.'));?></em>
+
+	<h3><?php p($l->t('User Instructions')); ?></h3>
```
Comment (nickvergessen) [inline templates/admin.php:70]:
```suggestion
	<h3><?php p($l->t('User instructions')); ?></h3>
```

File: templates/form/user.php [hunk 6]
```diff
@@ -15,6 +15,14 @@
 			<li><?php p($l->t('Welcome, you can create your account below.'));?></li>
 		</ul>
 		<?php } ?>
+
+		<?php if (!empty($_['additional_hint'])): ?>
+			<ul class="error">
```
Comment (nickvergessen) [inline templates/form/user.php:20]:
```suggestion
			<ul class="msg">
```

**Review by nickvergessen — CHANGES_REQUESTED** [review]
Looks good in general. Thanks for sending the PR!

The only thing hurting a bit is that the hints can not be translated. So when you have users with more than 1 language, they will all see the same language. But I guess it's okay for now.

Can you fix the minor comments? Then we can put this in. 😎

**Comment by pxlfrk** [comment]
> Looks good in general. Thanks for sending the PR!

Thanks @nickvergessen ! 👍 

> The only thing hurting a bit is that the hints can not be translated. So when you have users with more than 1 language, they will all see the same language. But I guess it's okay for now.

I totally understand your oint and also thought about this while adding the feature, but didn't have an alternative in mind. The only option to solve this right now would be to know which languages are in use and multiply the fields after that. But as the UI adapts to the system-language and the server doesn't know this (afaik) i don't see any chance of implementing this. I added a notice to the Settings-Panel that the configured text will be display to all users independent of their language.

> 
> Can you fix the minor comments? Then we can put this in. 😎
Sure, done that by writing this comment! 💯 

The php-cs check fails, but i'm not aware hot to solve them since php isn't my regular progamming language.

File: lib/Service/MailService.php [hunk 7]
```diff
@@ -102,6 +107,12 @@ public function sendTokenByMail(Registration $registration): void {
 			htmlspecialchars($body . ' ' . $this->l10n->t('Click the button below to continue.')),
 			$body
 		);
+		
+		// if the parameter is set through the settings panel add to body text
+		$email_verification_hint = $this->config->getAppValue('registration', 'email_verification_hint');
+ 		if (!empty($email_verification_hint)) {
+				$template->addBodyText($email_verification_hint);
+		};
```
Comment (nickvergessen) [inline lib/Service/MailService.php:115]:
```suggestion
		if (!empty($email_verification_hint)) {
			$template->addBodyText($email_verification_hint);
		};
```

This should fix the failing CS test

**Comment by nickvergessen** [comment]
I can take it from here. Thanks again for the PR!

**Comment by nickvergessen** [comment]
Merged via https://github.com/nextcloud/registration/commit/9e3b0b50e5cfa05aca13dca406f3184d3e2fda2e after fixing unit tests and adding new ones

**Comment by nickvergessen** [comment]
v0.6.0 released to the appstore

**Comment by pxlfrk** [comment]
@nickvergessen thx for shipping. I just updatet the App. What's the procedure for updating the translation files? If the community over at transifex translated the newly introduced strings, do they only ship within a new release of the app (e.g. 0.61)? Just wan't to make sure i get the process right 💯

**Comment by nickvergessen** [comment]
Oh right, should have waited a day or 2 before releasing. So yeah, will come with 0.6.1 I will check on friday if there was enough activity and otherwise after Xmas break in January.

**Pull request closed without merging**